### PR TITLE
User can add movie

### PR DESCRIPTION
**Pull request by p-duke**
@kellywu here's the tracking movie epic. Users can now add movies to their lists and remove items!

### CODE REVIEW

File: app/controllers/movies_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  def index
+    user = User.find(params[:user_id])
+    movies = user.movies
+    data = { movies: movies }
+    render json: data, status: 200
+  end
+
+  def create
+    @user = User.find(params[:user_id])
+    @movie = @user.movies.new(movie_params)
```
Comment (kellywu) [inline app/controllers/movies_controller.rb]: are you just returning json here? if so, these don't need to be instance variables

File: app/controllers/movies_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      data = { id: @movie.id, title: @movie.title, overview: @movie.overview, release_date: @movie.release_date, user_id: @movie.user_id }
+      render json: data
+    else
+      data = { errors: @movie.errors.messages }
+      render json: data, status: 422
+    end
+  end
+
+  def destroy
+    user = User.find(params[:user_id])
```
Comment (kellywu) [inline app/controllers/movies_controller.rb]: try doing this in 1 line with active record:  https://apidock.com/rails/ActiveRecord/QueryMethods/includes

File: app/javascript/packs/Main.jsx [hunk 3]
```diff
@@ -26,6 +27,10 @@ export default class Main extends React.Component {
           <Route exact path='/users/sign_out' render={() => (
              <LogOut />
           )} />
+
+          <Route path='/users/*' render={() => (
```
Comment (kellywu) [inline app/javascript/packs/Main.jsx]: any reason you're targeting all paths after /users/*?

File: app/javascript/packs/PopularMovies.jsx [hunk 4]
```diff
@@ -16,20 +16,23 @@ export default class PopularMovies extends React.Component {
...
-      data: {
-      },
-    }).then(function(response) {
-      self.context.store.dispatch({
-        type: 'GET_MOVIES',
-        data: response.data.results,
+    if (_.isEmpty(popularMovies)) {
+      axios({
+        method: 'GET',
+        url: 'https://api.themoviedb.org/3/discover/movie?api_key=8321a3cfc56c16d3d5e7144336d8a6e2&language=en-US&sort_by=popularity.desc&include_adult=false&include_video=false&page=1',
```
Comment (kellywu) [inline app/javascript/packs/PopularMovies.jsx]: consider extracting this URL into constants

File: app/javascript/packs/App.jsx [hunk 5]
```diff
@@ -5,9 +5,29 @@ import Main from './Main';
 export default class App extends React.Component {
   constructor() {
     super();
+    this.localRestore = this.localRestore.bind(this);
+  }
+
+  localRestore() {
```
Comment (kellywu) [inline app/javascript/packs/App.jsx]: cool, this is nice, and should probably be extracted out of `App`

File: app/javascript/packs/Header.jsx [hunk 6]
```diff
@@ -35,6 +35,10 @@ export default class Header extends React.Component {
               <li><Link to='/users/sign_up'>Sign Up</Link></li>
               <li><Link to='/users/sign_in'>Log In</Link></li>
               <li><Link to='/users/sign_out'>Log Out</Link></li>
+              { user
+                ? <li><Link to={'/users/'+user.id}>Profile</Link></li>
```
Comment (kellywu) [inline app/javascript/packs/Header.jsx]: i typically like to use string interpolation here `'/users/${user.id}`

File: app/javascript/packs/PopularMovies.jsx [hunk 7]
```diff
@@ -16,20 +16,23 @@ export default class PopularMovies extends React.Component {
...
+        data: {
+        },
+      }).then(function(response) {
+        self.context.store.dispatch({
+          type: 'GET_MOVIES',
+          data: response.data.results,
+        });
+      }).catch(function(error) {
+        console.log("Movie fetch blew up!", error);
```
Comment (kellywu) [inline app/javascript/packs/PopularMovies.jsx:32]: a nice feature would be showing the error on a UI

File: app/javascript/packs/Profile.jsx [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      self.setState({ errors: error.response.data })
+    });
+  }
+
+  render() {
+    const movies = this.state.movies;
+    const user = this.context.store.getState().user[0];
+
+    return (
+      <div className="container target">
```
Comment (kellywu) [inline app/javascript/packs/Profile.jsx:89]: a common pattern we like to use is having smart and dumb components

https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0

File: app/javascript/packs/SearchMovie.jsx [hunk 9]
```diff
@@ -32,33 +87,39 @@ export default class SearchMovie extends React.Component {
...
+          <form ref='form' action="/movies" method="post">
             <div className="form-group">
               <label>Search Movies</label>
               <input type="text" ref="query" name="query" className="form-control" placeholder="Try a search like 'star wars'" required />
             </div>
-            <button type="submit" className="btn btn-primary" onClick={this.search}>Search</button>
+            <div className="form-group">
+              <button type="submit" className="btn btn-primary" onClick={this.search}>Search</button>
+            </div>
+            { !_.isEmpty(movies)
```
Comment (kellywu) [inline app/javascript/packs/SearchMovie.jsx:104]: this is a bit hard to read, consider extracting into two methods.

e.g.

`noResults() { return <p>No Results</p>; }`
`results(movies) { ... }`

**Review by kellywu — COMMENTED** [review]
good stuff here!

File: app/javascript/packs/Main.jsx (hunk 3 above)
Comment (p-duke) [reply app/javascript/packs/Main.jsx]: yea this is definitely too broad of a match. Trying to match only as far as the users id for example '/users/:id'

**Pull request merged**